### PR TITLE
Version 0.8.1 build #114

### LINKED ISSUE #190: Strange client.properties issue that disconnected all people
```
Last packet IDs (Server->Client) before disconnection:
[('0x2', 47)]
Error in the Server->Client method:
Traceback (most recent call last):
  File "Wrapper.py/proxy.py", line 906, in handle
    if self.parse(id, original) and self.safe:
  File "Wrapper.py/proxy.py", line 853, in parse
    properties = client.properties
AttributeError: 'UUID' object has no attribute 'properties'
```

### PR DESCRIPTION
**Pull request by suresttexas00**
1) proxy packet Connect() -  fix proxys unable to connect () message to strict JSON. Turns out 1.9 is still using JSON for this field.  However, it is strict JSON with double quotes.  This now restores the original Red Bold that was in the previous 1.8 version. 
2) Remove an annoying debug message (that sends each player list packet to the console) that is not needed any longer.
3) Allow running the new 108 protocol (First Patch for 1.9.1 pre-release protocol 108)
4) Added ban-ip/pardon-ip code to proxy.py. 
5) Added a universal connect to allow 1.9 clients to connect to a 1.9.1 snapshot server.  There are some serious bugs with 1.9 server jar that are only fixed in the snapshots.
6) Added an item to prevent named OPs from being in the server player list ping.
7) Add the ipban commands to commands.py. 
8) Add the `import random` statement that was needed elsewhere by the random number command.  (commands.py …)
9) Correct problem with improperly capitalized names when someone does not capitalize a player name properly when running /perms users <IGN> info, causing the offline server to think it is a new player.
10) Correct problems with setGroup and hasGroup for new players.
11) Small syntax error in main.py `self.log.info("%s failed to load!" % plugin)` was "(plug)".
12) Small syntax error #270 in player.py line 69
13) in Main - refactor and make uuid code in main more consistent to deliver uuid _objects_ versus sometimes an object, sometimes a string, sometimes.. Something else (None, False..)
14) add more packets to mcpkt.py for use by player.py methods.
15) in Proxy.py - refactor some things for clarity and made uuid code more consistent to deliver uuid _objects_ versus sometimes an object, sometimes a string, ...
16) removing hard coded packet numbers from player.py code (make player methods more 1.9 compatible.
_17) introduced a typo that was undone in 19)._
_18) Undo same error as 17) introduced somewhere in 15) in proxy.py_
_19) fixed the error I introduced in 17) by removing `self.`_
20) Add error checking for missing "users" in various player permissions methods.
21) beefed up getUsername and added additional methods refreshUUID() and pollMojang().  after doing so and testing it, I have found that wrapper only uses this function in one place in Minecraft.py (getAllplayers()).. a sad under-usage that I think can be corrected.  The local wrapper usercache will now collect username history which includes all names, the current name and the original name.
I also changed the "lookupUsername" method (which actually looks up uuids) to simply "lookupuser" to make it less misleading.  That change was made in proxy.py and minecraft.py where applicable also...
22) proxy.py - changed more variable names and added some doc string info to clarify the actual functions of things, for instance formatUUID had arguments like "username" where the actual argument being passed is not a username, but a uuid that is being formatted.
23) commands.py - fixed the little bug that was introduced somewhere between 7-12) - messing up permissions data because a UUID object was not stringed before assigning `self.wrapper.proxy.lookupUser(username)` to a string value uuid for use in the permissions (and resulting in json errors during Storage.periodic.save() ).  I actually think this was always bad because the code already assumed a string coming in (which was not always the case).
24) api/Minecraft.py - nothing really, just a little line to remove "False" and "None" items from player directory.
25) api/Player.py - just "lookupUsername" changes to "lookupuser"
_26-28) correct player flying value typo, remove a debug print statement and add `self.` prefix to logs items._
29 - 33) Minecraft.py. player.py, --main--.py, commands.py, proxy.py:
Removed most duplicated uuid and user-caching from proxy and centralized to `__main__`.  Removed uuid handling code references to proxy out of player, minecraft, and commands.  Added small print debug line into commands.py to findout 'what' the 'command' in line 37 `self.commands[pluginID][command](payload["player"], payload["args"])` is supposed to be (it is not defined anywhere).  Discontinuing usage of the proxy 'user-cache' and using the wrappers `__main__py` self.usercache instead.. with a few extra fields to allow for local user name, name history, etc.
34-35) 	Missed a bad leftover proxy.lookupUUID reference in Commands.
36-42) Due to some permission parts being explicitly unicode and others not, it was necessary to str() the uuid to any call to lookupUsernamebyUUID().
43) Since Benbot @BenBaptist has not pulled this yet, we need to make this work up to the latest stable 1.9.2 release.
44) Typo in doc string corrected.
45) fix #190 (hopefully) getclientbyserveruuid() would not return a clear False if something was wrong (like could not find client).. so even thougth cleint object was not returned, code attempts to foll with it like it is a client object.

also added: 46+++ ... Huge refactor!

__Notable features__

- wrapper split to 5 folders, `__init__.py`, and `__main__.py`
- Proxy split into /proxy: base, server, client, packet, and mcpacket modules.
- API - same structure
- Management folder - now contains dashboard.py web.py, and HTML folder.
- Utils - encryption.py and log.py - added:  
    * termcolors.py (colored console logging!)  
    * six - PY3 <=> PY2 compatibility code  
    * helpers - place for helpful code snippets used regularly in wrapper code.  
- Core - contains most of what was originally the central wrapper base.

Logging system has been improved and now has an additional level TRACE for tracing packets through the proxy.  Logging has a different config file now (logging.json) and does not use the wrapper.properties setting.

Ban code is now handled by wrapper for both IP and UUID bans.
_**Wrapper uses the "expires" field to allow bans with a date to expire at that date/time!  "forever" bans (the default vanilla text)continue as always.  Expiration date must be written as "2016-04-14 22:04:53 -0400" (just as the vanilla server would write it)._
_**Wrapper writes ban information into the vanilla server banned-ips and banned-players files.. therefore, if the server is ever operated in online mode, the bans remain effective._

### CODE REVIEW

**Comment by suresttexas00** [comment]
I must close this for now until it is done

**Comment by JasonBristol** [comment]
+1, looks like you have been busy ;) Have been sick the past few days but this looks like a lot of our painpoints you have identified and resolved. Good to see the UUID issue has been identified, setGroup() has also been an annoyance. I will take a look at some of your other threads and see if there is anything I can help with

**Comment by suresttexas00** [comment]
Yes, I am not quite ready with this because I think there is one more error I need to track down that is feeding bad UUID information into the permissions system.  I'll re-open the PR when it is fixed.

I think (and I was considering asking) you could work on the main client threading portions.. see if you can make a thread-safe version that might close out possible memory leaks

**Comment by JasonBristol** [comment]
Awesome sounds good, testing some solutions for our the threading issues so I should have a PR somewhat soonish

**Comment by suresttexas00** [comment]
Those last five (5) commits now bring wrapper to a point that I am pretty happy with and even runs well under my rather burdensome plugin system :D !!  Here's a summary of the last 5:

Main.py - I beefed up getUsername and added additional methods refreshUUID() and pollMojang().  after doing so and testing it, I have found that wrapper only uses this function in one place in Minecraft.py (getAllplayers()).. a sad under-usage that I think can be corrected.  The local wrapper usercache will now collect username history which includes all names, the current name and the original name.
I also changed the "lookupUsername" method (which actually looks up uuids) to simply "lookupuser" to make it less misleading.  That change was made in proxy.py and minecraft.py where applicable also...

Proxy.py - changed more variable names and added some doc string info to clarify the actual functions of things, for instance formatUUID had arguments like "username" where the actual argument being passed is not a username, but a uuid that is being formatted.

Commands.py - fixed the little bugger that was holding up this PR - messing up permissions data because a UUID object was not stringed before assigning `self.wrapper.proxy.lookupUser(username)` to a string value uuid for use in the permissions (and resulting in json errors during Storage.periodic.save() ).

API/Minecraft.py - nothing really, just a little line to remove "False" and "None" items from player directory.

API/Player.py - just "lookupUsername" changes to "lookupuser"

**Comment by suresttexas00** [comment]
Gotta take this down one more time... UUID stuff needs some more refactoring and shoring up.

I am discontinuing the wrapper 'user-cache' and using the wrappers `__main__py` usercache (which is presently always 'empty' in proxy mode).  Why?  The usercache I am putting in wrapper is more fully featured than the simple UUID{ UUID: Name} dictionary that proxy.storage["user-cache"] is using.  Having two caches makes no sense and makes the code HARD to follow and debug.  The new wrapper.usercache has these fields;
```
self.usercache[useruuid] = {"time": time.time(),"original": None, "name": None,
										"online": True, "localname": None, "names": []}
```
- "localname" will be the local server name (for detecting name changes or maintaining a local nickname).  
- "names" is a list of previous names user has held in this format:

        `{"name": oldname, "date": changetime}`
- "online" will be the same as it always was (which is - "I don't know what").  True...
- "original" is the original name the player opened a minecraft account with
- "name" is the present valid name for the account.
- "time" - date of lookup 

..whew!

I'll re-open PR when done!

**Comment by suresttexas00** [comment]
While we are at it, I should add "IP" to the cache.

**Comment by suresttexas00** [comment]
I'm gonna stew over this a few days and see how it runs.   Then I'll bless it off and say it is ready to be pulled!

**Comment by suresttexas00** [comment]
Unless I find any bugs, all my further improvements and updates will be made on a new branch and a separate PR.

**Comment by suresttexas00** [comment]
Tonight was the first night of testing the permission coding thoroughly with all the nested groups and so forth that I have on my server..

**Comment by suresttexas00** [comment]
Added stable 1.9.2 protocol (108) to wrapper... no harm since Ben has not pulled yet.  Otherwise, I want this to be a stable version point.

**Comment by suresttexas00** [comment]
I am going to close this.  The new versions are a lot better and include all of this as their base anyway...  I'll be back next week to test for bugs.  If there are no problems I'll re-open this with the additional changes.
